File: packages/tools/src/tools/glob.test.ts
import { test } from "node:test";
import { strictEqual, ok } from "node:assert";
import { mkdir, rm, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { globTool } from "./glob.js";
import type { ToolContext } from "@forge/types";

const testDir = join(tmpdir(), `forge-glob-test-${Date.now()}`);

function makeCtx(cwd = testDir): ToolContext {
  return {
    cwd,
    sessionId: "test-session",
    threadId: "test-thread",
    signal: new AbortController().signal,
    emit: () => {},
  };
}

test("Glob: finds *.ts files", async () => {
  await mkdir(testDir, { recursive: true });
  await writeFile(join(testDir, "abed.ts"), "// Abed Nadir");
  await writeFile(join(testDir, "troy.ts"), "// Troy Barnes");
  await writeFile(join(testDir, "annie.js"), "// Annie Edison");

  const result = await globTool.handler({ pattern: "*.ts" }, makeCtx());

  strictEqual(result.isError, undefined);
  ok(result.content[0].type === "text");
  const text = result.content[0].text;
  ok(text.includes("abed.ts"));
  ok(text.includes("troy.ts"));
  ok(!text.includes("annie.js"));

  await rm(testDir, { recursive: true });
});

test("Glob: uses custom path", async () => {
  await mkdir(testDir, { recursive: true });
  const subdir = join(testDir, "study-room");
  await mkdir(subdir, { recursive: true });
  await writeFile(join(subdir, "britta.md"), "# Britta Perry");
  await writeFile(join(testDir, "dean.md"), "# Dean Pelton");

  const result = await globTool.handler(
    { pattern: "*.md", path: subdir },
    makeCtx(),
  );

  strictEqual(result.isError, undefined);
  ok(result.content[0].type === "text");
  const text = result.content[0].text;
  ok(text.includes("britta.md"));
  ok(!text.includes("dean.md"));

  await rm(testDir, { recursive: true });
});

test("Glob: no matches returns graceful message", async () => {
  await mkdir(testDir, { recursive: true });

  const result = await globTool.handler(
    { pattern: "*.nonexistent" },
    makeCtx(),
  );

  strictEqual(result.isError, undefined);
  ok(result.content[0].type === "text");
  strictEqual(result.content[0].text, "(no matches)");

  await rm(testDir, { recursive: true });
});

test("Glob: sorts by modification time (newest first)", async () => {
  await mkdir(testDir, { recursive: true });

  // Write files with slight delays to ensure different mtimes
  await writeFile(join(testDir, "old.txt"), "old");
  await new Promise((resolve) => setTimeout(resolve, 10));
  await writeFile(join(testDir, "new.txt"), "new");

  const result = await globTool.handler({ pattern: "*.txt" }, makeCtx());

  strictEqual(result.isError, undefined);
  ok(result.content[0].type === "text");
  const lines = result.content[0].text.split("\n");
  // Newest file should be first
  ok(lines[0].includes("new.txt"));
  ok(lines[1].includes("old.txt"));

  await rm(testDir, { recursive: true });
});

File: internal/tools/glob_test.go
package tools

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"github.com/jelmersnoeck/forge/internal/types"
	"github.com/stretchr/testify/require"
)

func TestGlobTool(t *testing.T) {
	tests := map[string]struct {
		setup func(t *testing.T, dir string) (map[string]any, types.ToolContext)
		want  func(*testing.T, types.ToolResult, error)
	}{
		"simple pattern": {
			setup: func(t *testing.T, dir string) (map[string]any, types.ToolContext) {
				_ = os.WriteFile(filepath.Join(dir, "troy.txt"), []byte("content"), 0644)
				_ = os.WriteFile(filepath.Join(dir, "abed.txt"), []byte("content"), 0644)
				_ = os.WriteFile(filepath.Join(dir, "jeff.go"), []byte("content"), 0644)

				return map[string]any{"pattern": "*.txt"}, types.ToolContext{CWD: dir}
			},
			want: func(t *testing.T, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)
				r.Contains(result.Content[0].Text, "troy.txt")
				r.Contains(result.Content[0].Text, "abed.txt")
				r.NotContains(result.Content[0].Text, "jeff.go")
			},
		},
		"recursive pattern": {
			setup: func(t *testing.T, dir string) (map[string]any, types.ToolContext) {
				_ = os.MkdirAll(filepath.Join(dir, "greendale", "library"), 0755)
				_ = os.WriteFile(filepath.Join(dir, "root.go"), []byte("content"), 0644)
				_ = os.WriteFile(filepath.Join(dir, "greendale", "study.go"), []byte("content"), 0644)
				_ = os.WriteFile(filepath.Join(dir, "greendale", "library", "books.go"), []byte("content"), 0644)

				return map[string]any{"pattern": "**/*.go"}, types.ToolContext{CWD: dir}
			},
			want: func(t *testing.T, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)
				r.Contains(result.Content[0].Text, "root.go")
				r.Contains(result.Content[0].Text, "study.go")
				r.Contains(result.Content[0].Text, "books.go")
			},
		},
		"no matches": {
			setup: func(t *testing.T, dir string) (map[string]any, types.ToolContext) {
				_ = os.WriteFile(filepath.Join(dir, "file.txt"), []byte("content"), 0644)
				return map[string]any{"pattern": "*.go"}, types.ToolContext{CWD: dir}
			},
			want: func(t *testing.T, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)
				r.Equal("(no matches)", result.Content[0].Text)
			},
		},
		"sorted by mtime": {
			setup: func(t *testing.T, dir string) (map[string]any, types.ToolContext) {
				// Create files with different modification times
				old := filepath.Join(dir, "old.txt")
				_ = os.WriteFile(old, []byte("old"), 0644)
				oldTime := time.Now().Add(-2 * time.Hour)
				_ = os.Chtimes(old, oldTime, oldTime)

				time.Sleep(10 * time.Millisecond)

				new := filepath.Join(dir, "new.txt")
				_ = os.WriteFile(new, []byte("new"), 0644)

				return map[string]any{"pattern": "*.txt"}, types.ToolContext{CWD: dir}
			},
			want: func(t *testing.T, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)
				lines := strings.Split(result.Content[0].Text, "\n")
				r.Len(lines, 2)
				// Newest first
				r.Contains(lines[0], "new.txt")
				r.Contains(lines[1], "old.txt")
			},
		},
		"custom path": {
			setup: func(t *testing.T, dir string) (map[string]any, types.ToolContext) {
				subdir := filepath.Join(dir, "subdir")
				_ = os.MkdirAll(subdir, 0755)
				_ = os.WriteFile(filepath.Join(subdir, "file.txt"), []byte("content"), 0644)

				return map[string]any{
					"pattern": "*.txt",
					"path":    subdir,
				}, types.ToolContext{CWD: dir}
			},
			want: func(t *testing.T, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)
				r.Contains(result.Content[0].Text, "file.txt")
			},
		},
		"skip directories": {
			setup: func(t *testing.T, dir string) (map[string]any, types.ToolContext) {
				_ = os.MkdirAll(filepath.Join(dir, "troy"), 0755)
				_ = os.WriteFile(filepath.Join(dir, "troy.txt"), []byte("content"), 0644)

				return map[string]any{"pattern": "troy*"}, types.ToolContext{CWD: dir}
			},
			want: func(t *testing.T, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)
				// Should only match the file, not the directory
				r.Equal("troy.txt", strings.TrimSpace(result.Content[0].Text))
			},
		},
		"complex pattern": {
			setup: func(t *testing.T, dir string) (map[string]any, types.ToolContext) {
				_ = os.MkdirAll(filepath.Join(dir, "src", "tools"), 0755)
				_ = os.MkdirAll(filepath.Join(dir, "pkg", "tools"), 0755)
				_ = os.WriteFile(filepath.Join(dir, "src", "tools", "read.go"), []byte("content"), 0644)
				_ = os.WriteFile(filepath.Join(dir, "pkg", "tools", "write.go"), []byte("content"), 0644)
				_ = os.WriteFile(filepath.Join(dir, "src", "main.go"), []byte("content"), 0644)

				return map[string]any{"pattern": "**/tools/*.go"}, types.ToolContext{CWD: dir}
			},
			want: func(t *testing.T, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)
				r.Contains(result.Content[0].Text, "read.go")
				r.Contains(result.Content[0].Text, "write.go")
				r.NotContains(result.Content[0].Text, "main.go")
			},
		},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			dir := t.TempDir()
			input, ctx := tc.setup(t, dir)
			tool := GlobTool()
			result, err := tool.Handler(input, ctx)
			tc.want(t, result, err)
		})
	}
}
